### PR TITLE
Enable `AppFocus` and `AppBlur` in terminal emulators

### PR DESCRIPTION
**Pull request by davep**
This enables support for receiving and handling FocusIn and FocusOut sequences, and turns then into `AppFocus` and `AppBlur` events. Support for enabling this is added to both the "Linux" and Windows drivers. From now on, when a terminal emulator window that supports it sends the FocusIn and FocusOut sequences the app will receive `AppFocus` and `AppBlur`.

An automatic effect of this now is that, when the window loses focus, focus will be removed from the widget that has focus. This was already an established feature of Textual applications running via textual-web; it just now extends it to the drivers used on GNU/Linux, macOS and Windows.

The `AppFocus` and `AppBlur` events could be useful too.

https://github.com/Textualize/textual/assets/28237/33bf3b9e-ab05-4c8c-89de-01da79cccace

It should be noted that this is technically a breaking change.

Implements #4263 (which in turn stems from #4259).

It also fixes a problem with the current implementation of `AppFocus` and `AppBlur` that is mentioned in a comment in #4263.

---

A natural question is going to be, how widely supported is this? So far the following have been tested and shown to support this:

- macOS
  - Kitty
  - iTerm2
  - Contour
  - terminal.app
  - WezTerm
  - rio
  - Tabby
  - VSCode Terminal
- GNU/Linux
  - GNOME Terminal
  - alacritty (via @TomJGooding)
  - urxvt (via @TomJGooding)
  - xterm
  - Konsole
- Windows 11
  - Terminal
  - Hyper
  - conemu (via @TomJGooding)

It has been noted by @TomJGooding that iTerm2 has a setting for turning this off[^1]; when turned off the enable/disable sequence appears to be swallowed and ignored.

It would appear that vterm under Emacs doesn't support this, but running my test app had the same effect as running it under iTerm with support turned off (that is, the enable/disable sequences seem to be swallowed and ignored).

**NOTE:** This is marked as and should be treated as a breaking change. This can potentially affect application styling in a way that would not have been an issue before. For example, before this change, this application would always look the same no matter if the terminal emulator it was within had focus or not:

```python
from textual.app import App, ComposeResult
from textual.widgets import Input

class OneWidgetApp(App[None]):

    CSS = """
    Screen {
        align: center middle;
    }

    Input:focus {
        border: round red;
        width: 50%;
        background: pink;
    }
    """

    def compose(self) -> ComposeResult:
        yield Input(placeholder="Please give your answer")

if __name__ == "__main__":
    OneWidgetApp().run()
```

from now on the layout of the application will look very different when focus moves away form the window.

[^1]: Because of course it does! What doesn't iTerm have a setting for?!?

### CODE REVIEW

File: src/textual/_xterm_parser.py [hunk 1]
```diff
@@ -20,6 +20,8 @@
 )
 _re_bracketed_paste_start = re.compile(r"^\x1b\[200~$")
 _re_bracketed_paste_end = re.compile(r"^\x1b\[201~$")
+_re_focusin = re.compile(r"^\x1b\[I$")
```
Comment (willmcgugan) [inline src/textual/_xterm_parser.py:23]: It doesn't look like this needs to be a regex. Could it not be a simple string?

Ditto for the bracketed paste stuff. Unless I'm missing something, it looks like they could also just be strings.

Comment (davep) [reply src/textual/_xterm_parser.py:23]: I was going to ask you that very question in the morning, oddly enough. Couldn't for the life of me see why the bracketed paste checking used a regex but followed the pattern with a mental note to ask why it was done that way (although looking at fe151a7f25cfd7f1134ebafbddc7eeade1c18ccb it seems I'd have been asking the wrong person anyway).

If there was no reason I'll make them simple strings; makes way more sense to me.

Comment (willmcgugan) [reply src/textual/_xterm_parser.py:23]: Probably just following a pattern. But yeah, let's go with strings. And if you wouldn't mind changing the bracketed paste ones while you are there...

Comment (davep) [reply src/textual/_xterm_parser.py:23]: Will do.

**Review by willmcgugan — APPROVED** [review]

**Comment by davep** [comment]
Adding for future reference if someone comes looking: this is also supported by `tmux` if `focus-events` is set to `on`.

**Pull request merged**